src/game_events/handlers.cpp

```diff
@@ -63,15 +110,210 @@ void event_handler::handle_event(const queued_event& event_info, game_lua_kernel
+	bool operator()(const queued_event&) const override
+	{
+		return ssf_.match(resources::controller->current_side());
+	}
+	void serialize(config& cfg) const override
+	{
+		cfg.add_child("filter_side", ssf_.get_config());
+	}
+private:
+	side_filter ssf_;
+};
+
+struct filter_unit : public event_filter {
+	filter_unit(const vconfig& cfg, bool first) : suf_(cfg.make_safe()), first_(first) {}
+	bool operator()(const queued_event& event_info) const override
+	{
+		const auto& loc = first_ ? event_info.loc1 : event_info.loc2;
+		auto unit = resources::gameboard->units().find(loc);
+		return loc.matches_unit_filter(unit, suf_);
+	}
+	void serialize(config& cfg) const override
+	{
+		cfg.add_child(first_ ? "filter" : "filter_second", suf_.to_config());
+	}
+private:
+	unit_filter suf_;
+	bool first_;
+};
+
+struct filter_attack : public event_filter {
+	filter_attack(const vconfig& cfg, bool first) : swf_(cfg.make_safe()), first_(first) {}
+	bool operator()(const queued_event& event_info) const override
+	{
+		const unit_map& units = resources::gameboard->units();
+		const auto& loc = first_ ? event_info.loc1 : event_info.loc2;
+		auto unit = units.find(loc);
+		if(unit != units.end() && loc.matches_unit(unit)) {
+			const config& attack = event_info.data.child(first_ ? "first" : "second");
+			return swf_.empty() || matches_special_filter(attack, swf_);
+		}
+		return false;
+	}
+	void serialize(config& cfg) const override
+	{
+		cfg.add_child(first_ ? "filter_attack" : "filter_second_attack", swf_.get_config());
+	}
+private:
+	vconfig swf_;
+	bool first_;
+};
+
+struct filter_formula : public event_filter {
+	filter_formula(const std::string& formula) : formula_(formula) {}
+	bool operator()(const queued_event& event_info) const override
+	{
+		wfl::gamestate_callable gs;
+		wfl::event_callable evt(event_info);
+		wfl::formula_callable_with_backup data(evt, gs);
+		return formula_.evaluate(data).as_bool();
+	}
+	void serialize(config& cfg) const override
+	{
+		std::string code = formula_.str();
+		if(cfg.has_attribute("filter_formula")) {
+			// This will probably never happen in practice, but handle it just in case it somehow can
+			code = "(" + cfg["filter_formula"].str() + ") and (" + code + ")";
+		}
+		cfg["filter_formula"] = code;
+	}
+private:
+	wfl::formula formula_;
+};
+
+struct filter_lua : public event_filter {
+	filter_lua(const std::string& code, const vconfig& args) : code_(code), args_(args)
+	{
+		event_ref_ = resources::lua_kernel->save_wml_event("filter", "", code);
+	}
+	bool operator()(const queued_event& event_info) const override
+	{
+		bool result;
+		return resources::lua_kernel->run_wml_event(event_ref_, args_, event_info, &result) && result;
+	}
+	void serialize(config& cfg) const override
+	{
+		if(cfg.has_attribute(("filter"))) {
+			WRN_NG << "Somehow an event has two filter= attributes";
+		}
+		cfg["filter"] = code_;
+	}
+	~filter_lua()
```

Comment:
Remember the Rule of Three, otherwise there are many ways to get `clear_wml_event` calls without corresponding `save_wml_event` calls.

Reply:
You mean having a move constructor, right? I think the existing use is safe due to guaranteed copy elision, but I guess it can't hurt to be extra certain.

I'm pretty sure an extra unnecessary `clear_wml_event` call on the same reference is safe, though; but if it you called `run_wml_event` on a reference that was already cleared, that might be bad.

Reply:
Mainly it's about the copy constructor and copy assignment. Rule of Three is that the compiler generates those for you unless you tell it not to, and so the destructor can run more often than the user-defined constructor does.

The move constructor and move assignment aren't automatically generated if there's a user-defined destructor, but for backward compatibility the copy versions still are.

Reply:
So then what you want me to do is delete the copy constructor?

Reply:
Both the copy constructor and the copy assignment.